File: 2015/Day05/Part01/rune.go
package main

import (
	"fmt"
	"strings"
)

func isVowel(ch rune) bool {
	return strings.ContainsRune("aeiou", ch)
}

func isNice(s string) bool {
	vowelCount, doubleLetter, containsBadStr := 0, false, false

	// Check for disallowed strings
	for _, badStr := range []string{"ab", "cd", "pq", "xy"} {
		if strings.Contains(s, badStr) {
			containsBadStr = true
			break
		}
	}

	// Check for vowels and double letters
	for i, ch := range s {
		if isVowel(ch) {
			vowelCount++
		}
		if i > 0 && ch == rune(s[i-1]) {
			doubleLetter = true
		}
	}

	return vowelCount >= 3 && doubleLetter && !containsBadStr
}

func main() {
	strings := []string{
		"uxcplgxnkwbdwhrp", "suerykeptdsutidb", "dmrtgdkaimrrwmej", "ztxhjwllrckhakut",
		"gdnzurjbbwmgayrg", "gjdzbtrcxwprtery", "fbuqqaatackrvemm", "pcjhsshoveaodyko",
		"lrpprussbesniilv", "mmsebhtqqjiqrusd", "vumllmrrdjgktmnb", "ptsqjcfbmgwdywgi",
		"mmppavyjgcfebgpl", "zexyxksqrqyonhui", "npulalteaztqqnrl", "mscqpccetkktaknl",
		"ydssjjlfejdxrztr", "jdygsbqimbxljuue", "ortsthjkmlonvgci", "jfjhsbxeorhgmstc",
		"vdrqdpojfuubjbbg", "xxxddetvrlpzsfpq", "zpjxvrmaorjpwegy", "laxrlkntrukjcswz",
		"pbqoungonelthcke", "niexeyzvrtrlgfzw", "zuetendekblknqng", "lyazavyoweyuvfye",
		"tegbldtkagfwlerf", "xckozymymezzarpy",
	}

	niceCount := 0
	for _, s := range strings {
		if isNice(s) {
			niceCount++
		}
	}

	fmt.Printf("Number of nice strings: %d\n", niceCount)
}

File: 2015/Day05/Part01/rune.kt
import java.io.File

// Проверка, является ли символ гласной
fun isVowel(ch: Char): Boolean {
    return ch in "aeiou"
}

// Проверка, является ли строка "nice" 
fun isNice(s: String): Boolean {
    var vowelCount = 0
    var doubleLetter = false
    var containsBadStr = false

    // Список недопустимых строк
    val badStrs = listOf("ab", "cd", "pq", "xy")
    for (bad in badStrs) {
        if (bad in s) {
            containsBadStr = true
            break
        }
    }

    // Проверка наличия гласных и повторяющихся букв
    for (i in s.indices) {
        if (isVowel(s[i])) {
            vowelCount++
        }
        if (i > 0 && s[i] == s[i - 1]) {
            doubleLetter = true
        }
    }

    // Возвращается true, если строка соответствует всем критериям
    return vowelCount >= 3 && doubleLetter && !containsBadStr
}

fun main(args: Array<String>) {
    val inputFile = File("input.txt")
    var niceCount = 0

    // Чтение каждой строки из файла и подсчет "nice" строк
    inputFile.forEachLine { line ->
        if (isNice(line)) {
            niceCount++
        }
    }

    // Вывод количества "nice" строк
    println("Количество хороших строк: $niceCount")
}
